Pull Request init

sefeoz opened the pull request.

sefeoz committed:
.

sefeoz committed:
Yazar table i olusturma ve baglama

Migrationi requestleri ve controlleri ayarladiktan sonra viewlari hallettim ve route ayarlarini yaptim. Datayi AI ya eklettim.

sefeoz committed:
Stores table ve bağlantıları

semihkeskindev commented on kutuphane/app/Models/Book.php:
```diff
@@ -6,9 +6,22 @@
-    protected $fillable = ['kitap_adi', 'yazar', 'ISBN', 'image'];
+    protected $fillable = ['kitap_adi', 'yazar_id', 'ISBN', 'image'];
+    
     public function favoritedBy()
```
methodların her yerde return type'ı olmalı

semihkeskindev commented on kutuphane/app/Http/Controllers/BookController.php:
```diff
@@ -13,32 +13,74 @@ class BookController extends Controller
+            return redirect()->back()->with('error', 'Hata: ' . $e->getMessage())->withInput();
+        }
     }
     public function destroy($id){
```
route model binding kullanmalısın

semihkeskindev commented on kutuphane/app/Http/Controllers/BookController.php:
```diff
@@ -13,32 +13,74 @@ class BookController extends Controller
             return redirect()->back()->with('error', 'Hata: ' . $e->getMessage())->withInput();
         }
     }
     public function edit($id){
```
route model binding

sefeoz committed:
Turkce kelimeleri ingilizceye cevirdim

Return typelari duzelttim. Book modelindeki degiskenlerin isimlerini duzelttim. Eski ad kullanilan her yerde yeni adini yazdim

sefeoz committed:
Author's View Fix

sefeoz committed:
Route Model binding fix

sefeoz committed:
Bulk import and test files added

semihkeskindev commented on kutuphane/app/Http/Controllers/AuthorController.php:
```diff
@@ -0,0 +1,52 @@
+    }
+    public function store(AuthorStoreRequest $request) : RedirectResponse
+    {
+        Author::create($request->all());
```
$request->all() dediğin zaman, body'de gönderilen tüm parametreler create methodunun içine verilir demek oluyor.

yani senin form içerisinde gönderdiğin değerler dışında ben 'semih' => 123 diye bir değer gönderirsem, authors tablosuna bu değeri de yazmaya çalışır ve patlar.

Bu yüzden $request->validated() kullanman daha doğru olur. En azından Request sınıfında validate ettiğin keyleri alırsın array içine sadece.

semihkeskindev replied:
aynı konu diğer kodlarda da geçerli. genel bir refactor yaparsın tüm kodlarında.

semihkeskindev commented on kutuphane/app/Http/Controllers/AuthorController.php:
```diff
@@ -0,0 +1,52 @@
+    }
+    public function search(Request $request) : View
+    {
+        $search = $request->input('search');
```
ya search değeri boş gönderilirse? değer en azından minimum 3 karakter olsa iyi olur gibi

semihkeskindev commented on kutuphane/app/Http/Controllers/BookController.php:
```diff
@@ -7,62 +7,120 @@
+                $stores = [];
+                foreach($request->stores as $store_id){
+                    $stores[$store_id] = [
+                        'price' => $request->input("prices.{$store_id}"),
```
requestten aldığın değerler niye stores değeri altında değil. prices, stock, is_active de stores değeri altında tutabilirsin ayrı tutmuşsun extra efor harcamışsın burada

semihkeskindev commented on kutuphane/app/Http/Controllers/BookController.php:
```diff
@@ -7,62 +7,120 @@
+                }
+                $book->stores()->sync($stores);
+            } else {
+                $book->stores()->detach();
```
bu ne işe yarıyor?

sefeoz replied:
pivot tabloyla senkronize ediyor.

semihkeskindev replied:
neyi senkronize ediyor? bana çok gereksiz gibi geldi. bu kadar else if de çok hoş değil early return ile çözemez miyiz?

semihkeskindev commented on kutuphane/app/Http/Controllers/BulkImportController.php:
```diff
@@ -0,0 +1,57 @@
+
+        $importHistory = ImportHistory::create([
+            'filename' => $originalName,
+            'status' => 'pending',
```
status leri bir enum dosyasına taşıyalım.

semihkeskindev commented on kutuphane/app/Http/Controllers/BulkImportController.php:
```diff
@@ -0,0 +1,57 @@
+            'total_records' => 0,
+            'processed_records' => 0,
+            'successful_records' => 0,
+            'failed_records' => 0,
```
bunları extra burada 0 diye tanımlamaya gerek olmasın. bu kolonların varsayılan değeri bu olsun.

semihkeskindev commented on kutuphane/app/Http/Controllers/BulkImportController.php:
```diff
@@ -0,0 +1,57 @@
+            'failed_records' => 0,
+        ]);
+
+        ImportAuthorsJob::dispatch($fullPath, $importHistory->id);
```
ImportAuthorListJob daha doğru olur. plural ifadeler çok doğru kabul edilmiyor.

semihkeskindev commented on kutuphane/app/Http/Requests/AuthorStoreRequest.php:
```diff
@@ -0,0 +1,29 @@
+     */
+    public function authorize(): bool
+    {
+        return auth()->check() && auth()->user()->role === 'admin';
```
ooo rol yapısı mı ekledik :D Roller de enum'a taşınsın. BU tarz statik string değerleri enuma veya constant olarak tutmalıyız ki değiştirmesi takibi temiz olsun.

semihkeskindev commented on kutuphane/app/Jobs/AuthorImportJob.php:
```diff
@@ -0,0 +1,125 @@
+    public $tries = 3;
+
+    public function __construct(
+        private array $authorData,
```
bu array olmasın. burada DTO kullanalım. array içinde ne geleceği muallak oluyor.

semihkeskindev commented on kutuphane/app/Jobs/AuthorImportJob.php:
```diff
@@ -0,0 +1,125 @@
+            
+            if (!$importHistory) {
+                Log::error("ImportHistory bulunamadı: {$this->importHistoryId}");
+                return;
```
bunu böyle yaparsan job success olur. exception fırlatmalısın. exception fırlatacağına da "ImportHistory::findOrFail" kullan

semihkeskindev commented on kutuphane/app/Models/Author.php:
```diff
@@ -0,0 +1,82 @@
+    {
+        return $this->hasMany(Book::class);
+    }
+    public static function createOrSkip(array $data)
```
bu methoda gerk yok aslında, modellerin zaten firstOrCreate methodu var.

semihkeskindev commented on kutuphane/app/Models/Author.php:
```diff
@@ -0,0 +1,82 @@
+    }
+    public static function isDuplicate(string $name): bool
+    {
+        return static::where('name', trim($name))->exists();
```
bunun için bir method yazmak garip olmuş ama okay

semihkeskindev commented on kutuphane/app/Models/Author.php:
```diff
@@ -0,0 +1,82 @@
+    {
+        return static::where('name', trim($name))->exists();
+    }
+    public static function bulkImport(array $authors): array
```
burada da DTO isterim. array type yerine DTO kullanalım.

semihkeskindev commented on kutuphane/app/Models/ImportHistory.php:
```diff
@@ -0,0 +1,19 @@
+{
+    protected $fillable = [
+        'filename',
+        'status',
```
enum cast

semihkeskindev commented on kutuphane/app/Models/ImportHistory.php:
```diff
@@ -0,0 +1,19 @@
+    protected $fillable = [
+        'filename',
+        'status',
+        'total_records',
+        'processed_records',
+        'successful_records',
+        'failed_records',
```
integer cast

semihkeskindev commented on kutuphane/database/migrations/2025_07_29_140253_create_books_table.php:
```diff
@@ -13,10 +13,10 @@ public function up(): void
             $table->id();
-            $table->string('kitap_adi');
-            $table->string('yazar');
+            $table->string('book_name');
```
book_name yerine isimlendirmeyi name yapalım. convention'a uygun olsun.

sefeoz committed:
Github Reviews

DTOs and enums created. Functions fixed.

sefeoz committed:
Import History

Enum cast and Integer cast

sefeoz committed:
Books table col name change

book_name to name change

sefeoz committed:
php-cs-fixer

sefeoz committed:
Mailpit and Bulk Import Refactor

Bulk import refactor. Used Chain of Responsibility. Mailpit Installation

semihkeskindev commented on kutuphane/.php-cs-fixer.cache:
```diff
@@ -0,0 +1 @@
+{"php":"8.4.10","version":"3.85.1:v3.85.1#2fb6d7f6c3398dca5786a1635b27405d73a417ba","indent":"    ","lineEnding":"\n","rules":{"binary_operator_spaces":{"default":"at_least_single_space"},"blank_line_after_opening_tag":true,"blank_line_between_import_groups":true,"blank_lines_before_namespace":true,"braces_position":{"allow_single_line_empty_anonymous_classes":true},"class_definition":{"inline_co ...
```
bu dosya .gitignore da olmalı repoya gitmemeli.

semihkeskindev commented on kutuphane/app/DTOs/AuthorImportData.php:
```diff
@@ -0,0 +1,36 @@
+
+namespace App\DTOs;
+
+class AuthorImportData implements \JsonSerializable
```
güzel yapmışsın, bu arada ayrıca laravel-data paketini de incelemeni tavsiye ederim güzeldir. but not required.

semihkeskindev commented on kutuphane/app/Http/Controllers/BulkImportController.php:
```diff
@@ -0,0 +1,54 @@
+
+    public function store(Request $request): RedirectResponse
+    {
+        $request->validate([
+            'file' => 'required|file|mimes:xlsx,xls,csv|max:2048',
+        ]);
```
request sınıfına taşımalısın

semihkeskindev commented on kutuphane/app/Http/Controllers/BulkImportController.php:
```diff
@@ -0,0 +1,54 @@
+        return view('bulk-import.history', compact('imports'));
+    }
+
+    public function show($id): View
```
route model binding

semihkeskindev commented on kutuphane/app/Imports/Authors/AuthorImportPipeline.php:
```diff
@@ -0,0 +1,28 @@
+{
+	public static function build(): HandlerInterface
+	{
+		$validate = new ValidateHandler();
+		$normalize = new NormalizeHandler();
+		$duplicate = new DuplicateCheckHandler();
+		$persist = new PersistHandler();
```
bunlar Dependecy Injection'a uygun değil. + bunun yerine laravel pipeline kullan

semihkeskindev commented on kutuphane/app/Jobs/AuthorImportJob.php:
```diff
@@ -0,0 +1,91 @@
+        $importHistory->increment('processed_records');
+        $importHistory->increment('failed_records');
+
+        $currentLog = $importHistory->error_log ?? '';
+        $newLog = $currentLog . "\n" . date('Y-m-d H:i:s') . ": " . $errorMessage;
+        $importHistory->update(['error_log' => trim($newLog)]);
+
+        $this->checkIfCompleted($importHistory);
```
error log örneği var mı? hangi satırda hangi hataları aldığını da belirtiyor musun tutuyor musun?

semihkeskindev commented on kutuphane/app/Jobs/ImportAuthorListJob.php:
```diff
@@ -0,0 +1,106 @@
+
+            $data = Excel::toArray([], $this->filePath);
+
+            if (empty($data) || empty($data[0])) {
```
böyle yapacağına $data array ise array filter ile boş satırları temizleyebilirsin. empty($data[0] bu kontrole gerek kalmaz.

semihkeskindev replied:
sonrasında count alır 0 dan küçük mü bakarsın.

semihkeskindev commented on kutuphane/app/Models/Author.php:
```diff
@@ -0,0 +1,67 @@
+        foreach ($authors as $authorData) {
+            try {
+                if (!$authorData instanceof AuthorImportData) {
+                    throw new \InvalidArgumentException('bulkImport yalnızca AuthorImportData dizisi kabul eder');
```
bu kontrol burada olmasa daha iyi, çünkü bu modelden bu method çağrılıyorsa doğru olmasını bekleriz. kendi yazdığmız kod doğru olsun :D  ama docblock olarak array içinde ne gelmesi gerektiğini belirtebilirsin.

semihkeskindev replied:
+ burada import işleminde eğer tek bir kayıt hatalı ise hiçbir kayıt import edilmemeli. 1 kayıt dahi validationdan geçmezse bir kayıt hata verirse hiçbir kayıt import edilmemiş olmalı.

bunun için ref: laravel beginTransaction
